pennylane/grouping/group_observables.py

```diff
@@ -237,18 +238,23 @@ def group_observables(observables, coefficients=None, grouping_type="qwc", metho
     if coefficients is None:
         return partitioned_paulis
 
-    partitioned_coeffs = [[0] * len(g) for g in partitioned_paulis]
+    partitioned_coeffs = [
+        qml.math.cast_like([0] * len(g), coefficients) for g in partitioned_paulis
+    ]
 
     observables = copy(observables)
-    coefficients = copy(coefficients)
     for i, partition in enumerate(partitioned_paulis):
-        for j, pauli_word in enumerate(partition):
+        indices = []
+        for pauli_word in partition:
+            # find index of this pauli word in original observables
             for observable in observables:
                 if are_identical_pauli_words(pauli_word, observable):
                     ind = observables.index(observable)
```

Comment:
I assume this logic (replacing the `pop` below) means that we avoid finding a Pauli word twice!